### PR TITLE
implement adding a base64 encoded file

### PR DESCRIPTION
**Pull request by fgreinus**
* also supports adding uri data scheme prefixed files

This is my first attempt to fix the issue #238.

Feedback very welcome as it is my first contribution to this project :).

### CODE REVIEW

File: src/HasMedia/HasMediaTrait.php [hunk 1]
```diff
@@ -94,6 +95,50 @@ public function addMediaFromUrl(string $url)
...
+     * @param string $data
+     *
+     * @throws InvalidBase64Data
+     * @throws \Spatie\MediaLibrary\Exceptions\FileCannotBeAdded
+     *
+     * @return \Spatie\MediaLibrary\FileAdder\FileAdder
+     */
+    public function addMediaFromBase64(string $data)
+    {
+        if (! is_string($data)) {
```
Comment (freekmurze) [inline src/HasMedia/HasMediaTrait.php]: The scalar type hint on this function will ensure that `$data` will always be a string. Remove this `if` block.

File: src/HasMedia/HasMediaTrait.php [hunk 2]
```diff
@@ -94,6 +95,50 @@ public function addMediaFromUrl(string $url)
...
+     * Add a base64 encoded file to the medialibrary.
+     *
+     * @param string $data
+     *
+     * @throws InvalidBase64Data
+     * @throws \Spatie\MediaLibrary\Exceptions\FileCannotBeAdded
+     *
+     * @return \Spatie\MediaLibrary\FileAdder\FileAdder
+     */
+    public function addMediaFromBase64(string $data)
```
Comment (freekmurze) [inline src/HasMedia/HasMediaTrait.php]: Rename `$data` to `$base64data`

File: src/HasMedia/HasMediaTrait.php [hunk 3]
```diff
@@ -94,6 +95,50 @@ public function addMediaFromUrl(string $url)
...
+
+        // decoding and then reeconding should not change the data
+        if (base64_encode(base64_decode($data)) !== $data) {
+            throw InvalidBase64Data::create();
+        }
+
+        $data = base64_decode($data);
+
+        // temporarily store the decoded data on the filesystem to be able to pass it to the fileAdder
+        $tmpFile = tempnam(sys_get_temp_dir(), 'media-library');
```
Comment (freekmurze) [inline src/HasMedia/HasMediaTrait.php]: Rename `media-library` to `medialibrary` (that's a bit more consistent with the reset of the code)

File: tests/FileAdder/IntegrationTest.php [hunk 4]
```diff
@@ -336,4 +337,43 @@ public function it_will_throw_an_exception_when_adding_a_file_to_a_non_existing_
...
+            ->addMediaFromBase64($testBase64Data)
+            ->toMediaLibrary();
+
+        $this->assertFileExists($this->getMediaDirectory($media->id.'/'.$media->file_name));
+    }
+
+    /** @test */
+    public function it_will_throw_an_exception_when_adding_invalid_base64_data()
+    {
+        $this->expectException(InvalidBase64Data::class);
```
Comment (freekmurze) [inline tests/FileAdder/IntegrationTest.php]: Move this line to right before the line starting with ` $this->testModel`

File: src/HasMedia/HasMediaTrait.php (hunk 2 above)
Comment (fgreinus) [inline src/HasMedia/HasMediaTrait.php]: $base64data or $base64Data (the second seems to be more correct in favour of camel-case)

File: src/HasMedia/HasMediaTrait.php (hunk 1 above)
Comment (fgreinus) [inline src/HasMedia/HasMediaTrait.php]: First time actually using scalar type hints.. so I did not think about this neat behaviour.. thanks!

File: src/HasMedia/HasMediaTrait.php (hunk 2 above)
Comment (freekmurze) [inline src/HasMedia/HasMediaTrait.php]: Let's go for `$base64data`. That feels slightly more natural to me.

**Comment by fgreinus** [comment]
I fixed the things you noted in your comments. The failed travis builds seem to be related to an issue within the provisioning.

**Comment by freekmurze** [comment]
Yup, I just restarted the build and now it works fine.

**Comment by freekmurze** [comment]
Thank you very much for this PR.

**Comment by nelson6e65** [comment]
How can I get the URL? Files are stored without extensions when used `addMediaFromBase64()` :sweat_smile:

**Comment by saqueib** [comment]
@nelson6e65 did you manage to get the file extension, I am having same issue, image is uploaded but it missing the extension.

**Comment by nelson6e65** [comment]
I don't remember, @saqueib. :sweat_smile: 

I was able to display the image anyway directly, so I forget about it. Maybe later I'll check that project.

But... Maybe you can set the extension by using the ```->usingFileName(string)``` method. You can check the string for `'data:image/{jpeg|png|svg|etc};base64'` to get the extension and pass it to the method with a name.

**Comment by saqueib** [comment]
@nelson6e65 thanks, I also used `->usingFileName(string)` to get around this. thanks

**Comment by reppair** [comment]
Nice one!

**Pull request merged**